### ISSUE TITLE
non-negligible output error between Quantized Onnx Model (w/ QDQ nodes)  and converted TRT Int8 Model 

### ISSUE THREAD

**Issue by FreemanHsu** [NVIDIA/TensorRT#2103]
## Description
Hello，I'm trying to convert a quantized onnx model (with QDQ nodes) and convert it to tensorrt int8 model.
The converted trt int8 model, however, has a huge performance drop, the outputs of onnx and trt model have a big difference.
I then tried to quantize standard models and then convert to tensorrt to verify I did the convertion in a right way.
I used 
mnist (https://github.com/onnx/models/blob/main/vision/classification/mnist/model/mnist-12.onnx) and 
resnet50v1 (https://github.com/onnx/models/blob/main/vision/classification/resnet/model/resnet50-v1-12.onnx)

Corresponding quantized model:
[mnist quantized onnx](https://drive.google.com/file/d/133fhdfYUltfCby6FBr_ilqmVCtCHtf62/view?usp=sharing) 
[resnet quantized onnx](https://drive.google.com/file/d/1399bSAY49fsIzte8f_TYGdUlK-Y14H6D/view?usp=sharing)

The results,
1. The outputs of mnist quantized onnx and trt int8 model are very close (with error ~ 1e-7)
2. The outputs of resnet quantized onnx and trt int8 model are different (with max error ~0.4)

I wonder if I did anything wrong during model conversion, or should I expect quantized onnx model and trt int8 model to produce same (or very close) outputs in big models?

Thank you!

## Environment

**TensorRT Version**: 8.2.5.1
**NVIDIA GPU**: RTX 2080Ti
**NVIDIA Driver Version**: 460.73.01
**CUDA Version**: 10.2
**CUDNN Version**: 8.4.0
**Operating System**: Ubuntu 18.04
**Python Version (if applicable)**: 3.8.13
**PyTorch Version (if applicable)**: 1.11.0
**Baremetal or Container (if so, version)**: 22.05-py3 (to use trtexec and polygraphy)

## Relevant Files
[mnist quantized onnx](https://drive.google.com/file/d/133fhdfYUltfCby6FBr_ilqmVCtCHtf62/view?usp=sharing) 
[resnet quantized onnx](https://drive.google.com/file/d/1399bSAY49fsIzte8f_TYGdUlK-Y14H6D/view?usp=sharing)

## Steps To Reproduce

1. Quantized Onnx To TRT Int8 with python
```python
import tensorrt as trt

TRT_LOGGER = trt.Logger(trt.Logger.INFO)

network_flags = 1 << int(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)
network_flags = network_flags | (1 << int(trt.NetworkDefinitionCreationFlag.EXPLICIT_PRECISION))

file_path = '/path/to/quantized.onnx'
input_shapes = {"input": [(1, 3, 224, 224), (1, 3, 224, 224), (1, 3, 224, 224)]}

def build_profile(builder, network, profile_shapes, default_shape_value=1):
    def is_dimension_dynamic(dim):
        return dim is None or dim <= 0

    def override_shape(shape):
        return tuple([1 if is_dimension_dynamic(dim) else dim for dim in shape])

    profile = builder.create_optimization_profile()
    for idx in range(network.num_inputs):
        inp = network.get_input(idx)

        def get_profile_shape(name):
            if name not in profile_shapes:
                return None
            shapes = profile_shapes[name]
            if not isinstance(shapes, list) or len(shapes) != 3:
                G_LOGGER.critical("Profile values must be a list containing exactly 3 shapes (tuples or Dims), but received shapes: {:} for input: {:}.\nNote: profile was: {:}.\nNote: Network inputs were: {:}".format(shapes, name, profile_shapes, get_network_inputs(network)))
            return shapes

        if inp.is_shape_tensor:
            shapes = get_profile_shape(inp.name)
            if not shapes:
                rank = inp.shape[0]
                shapes = [(default_shape_value, ) * rank] * 3
                print("Setting shape input to {:}. If this is incorrect, for shape input: {:}, please provide tuples for min, opt, and max shapes containing {:} elements".format(shapes[0], inp.name, rank))
            min, opt, max = shapes
            profile.set_shape_input(inp.name, min, opt, max)
            print("Setting shape input: {:} values to min: {:}, opt: {:}, max: {:}".format(inp.name, min, opt, max))
        elif -1 in inp.shape:
            shapes = get_profile_shape(inp.name)
            if not shapes:
                shapes = [override_shape(inp.shape)] * 3
                print("Overriding dynamic input shape {:} to {:}. If this is incorrect, for input tensor: {:}, please provide tuples for min, opt, and max shapes containing values: {:} with dynamic dimensions replaced,".format(inp.shape, shapes[0], inp.name, inp.shape))
            min, opt, max = shapes
            profile.set_shape(inp.name, min, opt, max)
            print("Setting input: {:} shape to min: {:}, opt: {:}, max: {:}".format(inp.name, min, opt, max))
    if not profile:
        print("Profile is not valid, please provide profile data. Note: profile was: {:}".format(profile_shapes))
    return profile

with trt.Builder(TRT_LOGGER) as builder, builder.create_network(flags=network_flags) as network, trt.OnnxParser(network, TRT_LOGGER) as parser:
  with open(file_path, 'rb') as model:
    if not parser.parse(model.read()):
      print ('ERROR: Failed to parse the ONNX file.')
      for error in range(parser.num_errors):
        print (parser.get_error(error))

  config = builder.create_builder_config()
  config.max_workspace_size = 2 << 30
  config.flags = config.flags | 1 << int(trt.BuilderFlag.INT8)
  config.add_optimization_profile(build_profile(builder,
                                                network,
                                                profile_shapes=input_shapes))

  trt_engine = builder.build_engine(network, config)

engine_file = file_path.replace('.onnx', '.engine')
with open(engine_file, 'wb') as file:
  file.write(trt_engine.serialize())

```

2. I have also tried to convert using trtexec
```bash
trtexec --onnx=/path/to/quantized.onnx --int8 --saveEngine=/path/to/quantized.engine
```
the trtexec converted model can produce the same result as the model converted by python script above

3. Check outputs
use onnxruntime and python tensorrt lib respectively to do inference and then check output

4. I've also tried polygraphy tool, 
```bash
polygraphy run --onnxrt --trt --int8 --workspace=1G /path/to/resnet_quantized.onnx --onnx-outputs=resnetv17_dense0_fwd --trt-outputs=resnetv17_dense0_fwd
```
and it failed the check, the outputs look like this:

```
[I] Accuracy Comparison | onnxrt-runner-N0-06/29/22-12:09:12 vs. trt-runner-N0-06/29/22-12:09:12
[I]     Comparing Output: 'resnetv17_dense0_fwd' (dtype=float32, shape=(1, 1000)) with 'resnetv17_dense0_fwd' (dtype=float32, shape=(1, 1000))
[I]     Tolerance: [abs=1e-05, rel=1e-05] | Checking elemwise error
[I]         onnxrt-runner-N0-06/29/22-12:09:12: resnetv17_dense0_fwd | Stats: mean=5.072e-05, std-dev=1.639, var=2.6862, median=-0.083648, min=-4.1459 at (0, 865), max=6.1207 at (0, 111), avg-magnitude=1.3031
[I]             ---- Histogram ----
                Bin Range          |  Num Elems | Visualization
                (-4.15  , -3.12  ) |         11 | #
                (-3.12  , -2.09  ) |         82 | #############
                (-2.09  , -1.07  ) |        177 | #############################
                (-1.07  , -0.0393) |        240 | ########################################
                (-0.0393, 0.987  ) |        233 | ######################################
                (0.987  , 2.01   ) |        140 | #######################
                (2.01   , 3.04   ) |         75 | ############
                (3.04   , 4.07   ) |         28 | ####
                (4.07   , 5.09   ) |          9 | #
                (5.09   , 6.12   ) |          5 | 
[I]         trt-runner-N0-06/29/22-12:09:12: resnetv17_dense0_fwd | Stats: mean=5.0125e-05, std-dev=1.6183, var=2.6189, median=-0.079418, min=-4.0695 at (0, 865), max=5.9674 at (0, 111), avg-magnitude=1.2864
[I]             ---- Histogram ----
                Bin Range          |  Num Elems | Visualization
                (-4.15  , -3.12  ) |         11 | #
                (-3.12  , -2.09  ) |         79 | ############
                (-2.09  , -1.07  ) |        178 | #############################
                (-1.07  , -0.0393) |        244 | ########################################
                (-0.0393, 0.987  ) |        233 | ######################################
                (0.987  , 2.01   ) |        141 | #######################
                (2.01   , 3.04   ) |         73 | ###########
                (3.04   , 4.07   ) |         28 | ####
                (4.07   , 5.09   ) |         10 | #
                (5.09   , 6.12   ) |          3 | 
[I]         Error Metrics: resnetv17_dense0_fwd
[I]             Minimum Required Tolerance: elemwise error | [abs=0.18149] OR [rel=26.199] (requirements may be lower if both abs/rel tolerances are set)
[I]             Absolute Difference | Stats: mean=0.042742, std-dev=0.031624, var=0.0010001, median=0.036902, min=0.00013089 at (0, 512), max=0.18149 at (0, 434), avg-magnitude=0.042742
[I]                 ---- Histogram ----
                    Bin Range          |  Num Elems | Visualization
                    (0.000131, 0.0183) |        263 | ########################################
                    (0.0183  , 0.0364) |        232 | ###################################
                    (0.0364  , 0.0545) |        189 | ############################
                    (0.0545  , 0.0727) |        138 | ####################
                    (0.0727  , 0.0908) |         95 | ##############
                    (0.0908  , 0.109 ) |         53 | ########
                    (0.109   , 0.127 ) |         15 | ##
                    (0.127   , 0.145 ) |          9 | #
                    (0.145   , 0.163 ) |          5 | 
                    (0.163   , 0.181 ) |          1 | 
[I]             Relative Difference | Stats: mean=0.16496, std-dev=1.1332, var=1.2842, median=0.034209, min=4.4486e-05 at (0, 512), max=26.199 at (0, 359), avg-magnitude=0.16496
[I]                 ---- Histogram ----
                    Bin Range        |  Num Elems | Visualization
                    (4.45e-05, 2.62) |        993 | ########################################
                    (2.62    , 5.24) |          2 | 
                    (5.24    , 7.86) |          0 | 
                    (7.86    , 10.5) |          1 | 
                    (10.5    , 13.1) |          2 | 
                    (13.1    , 15.7) |          1 | 
                    (15.7    , 18.3) |          0 | 
                    (18.3    , 21  ) |          0 | 
                    (21      , 23.6) |          0 | 
                    (23.6    , 26.2) |          1 | 
[E]         FAILED | Difference exceeds tolerance (rel=1e-05, abs=1e-05)
[E]     FAILED | Mismatched outputs: ['resnetv17_dense0_fwd']

```

**Comment by zerollzeng** [comment]
Absolute Difference or Relative Difference is not the metric to evaluate a model's accuracy, the metric should be the MAP score in the validation set. 
in your case, the model is a classification model, both the onnx(max=6.1207 at (0, 111)) and TRT(max=5.9674 at (0, 111)) have the same max value position, which means both of them to produce the same classification result.

**Comment by ttyio** [comment]
closing since no activity for more than 3 weeks, please reopen if you still have question, thanks!

**Comment by zt706** [comment]
> Absolute Difference or Relative Difference is not the metric to evaluate a model's accuracy, the metric should be the MAP score in the validation set. in your case, the model is a classification model, both the onnx(max=6.1207 at (0, 111)) and TRT(max=5.9674 at (0, 111)) have the same max value position, which means both of them to produce the same classification result.

We have encountered the same problem, the output accuracy is very different after exporting from the QDQ onnx model to the trt engine model. And I am a super-resolution model, the result can be aligned with trt+fp16 during QDQ+onnxruntime , but QDQ onnx is exported to trt 
(use the **ppq export tool**: https://github.com/openppl-public/ppq/blob/master/ppq/utils/TensorRTUtil.py#L212), the result is abnormal
